File: sdk/python/chromatindb/_envelope.py
"""PQ envelope encryption for chromatindb SDK.

Implements multi-recipient envelope encryption using the KEM-then-Wrap pattern:
- Random DEK encrypts data once with ChaCha20-Poly1305
- Per-recipient: ML-KEM-1024 encap -> HKDF KEK -> AEAD-wrap DEK
- Versioned binary format: [magic:4][version:1][suite:1][count:2 BE][nonce:12][N x stanza:1648][ciphertext+tag]

Internal module -- import from chromatindb instead.
"""

from __future__ import annotations

import secrets
import struct
from typing import TYPE_CHECKING

import brotli
import oqs

from chromatindb.crypto import AEAD_TAG_SIZE, aead_decrypt, aead_encrypt, sha3_256
from chromatindb._hkdf import hkdf_derive
from chromatindb.exceptions import (
    DecompressionError,
    DecryptionError,
    MalformedEnvelopeError,
    NotARecipientError,
)

if TYPE_CHECKING:
    from chromatindb.identity import Identity

# Envelope format constants (per D-01 through D-08)
ENVELOPE_MAGIC = b"CENV"                     # 0x43454E56 (per D-01)
ENVELOPE_VERSION = 0x01                       # per D-02
CIPHER_SUITE_ML_KEM_CHACHA = 0x01            # per D-03
_STANZA_SIZE = 1648                           # 32 + 1568 + 48 (per D-06)
_FIXED_HEADER_SIZE = 20                       # 4 + 1 + 1 + 2 + 12
_KEM_CT_SIZE = 1568                           # ML-KEM-1024 ciphertext
_KEM_PK_HASH_SIZE = 32                        # SHA3-256 of KEM pubkey
_WRAPPED_DEK_SIZE = 48                        # 32-byte DEK + 16-byte AEAD tag
_MAX_RECIPIENTS = 256                         # per D-04
_HKDF_LABEL = b"chromatindb-envelope-kek-v1"  # per D-06/D-11
_ZERO_NONCE = b"\x00" * 12                    # per D-09

# Brotli compression constants (Phase 87)
CIPHER_SUITE_ML_KEM_CHACHA_BROTLI = 0x02   # ML-KEM-1024 + ChaCha20-Poly1305 + Brotli (per D-02)
_COMPRESS_THRESHOLD = 256                    # Skip compression below this size (per D-05)
_BROTLI_QUALITY = 6                          # Brotli quality level (per D-04)
MAX_DECOMPRESSED_SIZE = 100 * 1024 * 1024    # 100 MiB, matches MAX_BLOB_DATA_SIZE (per D-09)


def _safe_decompress(compressed: bytes, max_size: int) -> bytes:
    """Decompress Brotli data with output size cap.

    Uses streaming Decompressor to avoid allocating full output
    before checking size. Raises DecompressionError if output
    exceeds max_size.
    """
    dec = brotli.Decompressor()
    chunks: list[bytes] = []
    total = 0

    result = dec.process(compressed)
    total += len(result)
    if total > max_size:
        raise DecompressionError(
            f"Decompressed size {total} exceeds maximum {max_size}"
        )
    chunks.append(result)

    while not dec.is_finished():
        result = dec.process(b"")
        total += len(result)
        if total > max_size:
            raise DecompressionError(
                f"Decompressed size {total} exceeds maximum {max_size}"
            )
        chunks.append(result)

    return b"".join(chunks)


def envelope_encrypt(
    plaintext: bytes,
    recipients: list[Identity],
    sender: Identity,
    compress: bool = True,
) -> bytes:
    """Encrypt data for multiple PQ recipients using KEM-then-Wrap.

    The sender is always auto-included as a recipient (cannot lock yourself
    out). Duplicate recipients (by KEM public key hash) are deduplicated.
    Stanzas are sorted by pk_hash for O(log N) lookup during decryption.

    Args:
        plaintext: Arbitrary data to encrypt (may be empty).
        recipients: List of recipient identities (each must have KEM pubkey).
        sender: Sender identity (must have KEM keypair).

    Returns:
        Binary envelope: header + per-recipient stanzas + ciphertext.

    Raises:
        ValueError: If sender lacks KEM, any recipient lacks KEM, or
            too many recipients (> 256).
    """
    if not sender.has_kem:
        raise ValueError("Sender must have KEM keypair")

    # Build unique recipient set: sender first, then others (per D-15, D-17)
    all_recipients = [sender] + list(recipients)
    seen_hashes: dict[bytes, Identity] = {}
    for r in all_recipients:
        if not r.has_kem:
            raise ValueError("All recipients must have KEM public keys")
        pk_hash = sha3_256(r.kem_public_key)
        if pk_hash not in seen_hashes:
            seen_hashes[pk_hash] = r
    unique_recipients = list(seen_hashes.values())

    # Enforce recipient cap (per D-04)
    if len(unique_recipients) > _MAX_RECIPIENTS:
        raise ValueError(
            f"Too many recipients: {len(unique_recipients)} > {_MAX_RECIPIENTS}"
        )

    # Generate random DEK and data nonce (per D-05)
    dek = secrets.token_bytes(32)
    data_nonce = secrets.token_bytes(12)

    # Pass 1: KEM encapsulation for each recipient
    encap_results: list[tuple[bytes, bytes, bytes]] = []  # (pk_hash, kem_ct, kem_ss)
    for recipient in unique_recipients:
        kem = oqs.KeyEncapsulation("ML-KEM-1024")
        kem_ct, kem_ss = kem.encap_secret(recipient.kem_public_key)
        pk_hash = sha3_256(recipient.kem_public_key)
        encap_results.append((pk_hash, bytes(kem_ct), bytes(kem_ss)))

    # Sort by pk_hash for O(log N) binary search during decrypt (per D-07)
    encap_results.sort(key=lambda t: t[0])

    # Determine compression and suite (per D-01, D-02, D-05, D-06, D-07)
    actual_suite = CIPHER_SUITE_ML_KEM_CHACHA
    data_to_encrypt = plaintext

    if compress and len(plaintext) >= _COMPRESS_THRESHOLD:
        compressed = brotli.compress(plaintext, quality=_BROTLI_QUALITY)
        if len(compressed) < len(plaintext):
            data_to_encrypt = compressed
            actual_suite = CIPHER_SUITE_ML_KEM_CHACHA_BROTLI

    # Build partial header and wrap AD (per D-10)
    recipient_count = len(encap_results)
    partial_header = bytearray()
    partial_header.extend(ENVELOPE_MAGIC)
    partial_header.extend(struct.pack("B", ENVELOPE_VERSION))
    partial_header.extend(struct.pack("B", actual_suite))
    partial_header.extend(struct.pack(">H", recipient_count))
    partial_header.extend(data_nonce)

    # Wrap AD = partial header + all (pk_hash + kem_ct) pairs
    wrap_ad = bytearray(partial_header)
    for pk_hash, kem_ct, _ in encap_results:
        wrap_ad.extend(pk_hash)
        wrap_ad.extend(kem_ct)
    wrap_ad = bytes(wrap_ad)

    # Pass 2: Wrap DEK for each recipient (per D-09, D-11)
    stanzas: list[tuple[bytes, bytes, bytes]] = []  # (pk_hash, kem_ct, wrapped_dek)
    for pk_hash, kem_ct, kem_ss in encap_results:
        kek = hkdf_derive(salt=b"", ikm=kem_ss, info=_HKDF_LABEL, length=32)
        # Zero nonce is safe: KEM shared secret (and thus KEK) is unique
        # per encapsulation -- the key is never reused.
        wrapped_dek = aead_encrypt(dek, ad=wrap_ad, nonce=_ZERO_NONCE, key=kek)
        stanzas.append((pk_hash, kem_ct, wrapped_dek))

    # Build full header (per D-08)
    full_header = bytearray(partial_header)
    for pk_hash, kem_ct, wrapped_dek in stanzas:
        full_header.extend(pk_hash)
        full_header.extend(kem_ct)
        full_header.extend(wrapped_dek)
    full_header = bytes(full_header)

    # Encrypt data with full header as AD (per D-04, ENV-01)
    ciphertext = aead_encrypt(data_to_encrypt, ad=full_header, nonce=data_nonce, key=dek)

    return full_header + ciphertext


def envelope_decrypt(data: bytes, identity: Identity) -> bytes:
    """Decrypt a PQ envelope for the given identity.

    Finds the recipient stanza by matching stanza pk_hashes against the
    identity's KEM key ring map (supports decryption with any historical
    KEM key after rotation). Decapsulates the KEM ciphertext, derives
    KEK via HKDF, unwraps DEK, and decrypts the payload.

    Args:
        data: Complete binary envelope.
        identity: Identity with KEM secret key for decryption.

    Returns:
        Decrypted plaintext bytes.

    Raises:
        MalformedEnvelopeError: If envelope has invalid magic, version,
            cipher suite, zero recipients, or is truncated.
        NotARecipientError: If identity is not in recipient list or has
            no KEM keypair.
        DecryptionError: If AEAD authentication fails (tampered envelope).
    """
    # Validate minimum size
    if len(data) < _FIXED_HEADER_SIZE + _STANZA_SIZE + AEAD_TAG_SIZE:
        raise MalformedEnvelopeError("Envelope too short")

    # Parse fixed header
    magic = data[:4]
    if magic != ENVELOPE_MAGIC:
        raise MalformedEnvelopeError(
            f"Invalid magic: expected CENV, got {magic!r}"
        )

    version = data[4]
    if version != ENVELOPE_VERSION:
        raise MalformedEnvelopeError(f"Unsupported envelope version: {version}")

    suite = data[5]
    if suite not in (CIPHER_SUITE_ML_KEM_CHACHA, CIPHER_SUITE_ML_KEM_CHACHA_BROTLI):
        raise MalformedEnvelopeError(f"Unsupported cipher suite: {suite}")

    recipient_count = struct.unpack(">H", data[6:8])[0]
    if recipient_count == 0:
        raise MalformedEnvelopeError("Envelope has zero recipients")

    data_nonce = data[8:20]

    # Validate total size
    expected_header_size = _FIXED_HEADER_SIZE + recipient_count * _STANZA_SIZE
    if len(data) < expected_header_size + AEAD_TAG_SIZE:
        raise MalformedEnvelopeError("Envelope truncated")

    # Parse stanza pk_hashes for binary search
    stanza_offset = _FIXED_HEADER_SIZE
    pk_hashes: list[bytes] = []
    for i in range(recipient_count):
        offset = stanza_offset + i * _STANZA_SIZE
        pk_hashes.append(data[offset : offset + _KEM_PK_HASH_SIZE])

    # Find recipient stanza via key ring map lookup (per D-08)
    if not identity.has_kem:
        raise NotARecipientError("Identity has no KEM keypair for decryption")

    ring_map = identity._build_kem_ring_map()
    if not ring_map:
        raise NotARecipientError("Identity has no KEM secret key for decryption")

    # Scan stanza pk_hashes against the ring map
    matched_idx = None
    matched_kem = None
    for i, pk_hash in enumerate(pk_hashes):
        if pk_hash in ring_map:
            matched_idx = i
            matched_kem = ring_map[pk_hash]
            break

    if matched_idx is None:
        raise NotARecipientError("Identity not in envelope recipient list")

    # Extract stanza fields using matched index
    stanza_base = stanza_offset + matched_idx * _STANZA_SIZE
    kem_ct = data[stanza_base + _KEM_PK_HASH_SIZE : stanza_base + _KEM_PK_HASH_SIZE + _KEM_CT_SIZE]
    wrapped_dek = data[stanza_base + _KEM_PK_HASH_SIZE + _KEM_CT_SIZE : stanza_base + _STANZA_SIZE]

    # Decapsulate with matched key from ring
    kem_ss = bytes(matched_kem.decap_secret(kem_ct))

    # Derive KEK and unwrap DEK (per D-09, D-11)
    kek = hkdf_derive(salt=b"", ikm=kem_ss, info=_HKDF_LABEL, length=32)

    # Build wrap AD: partial header + all (pk_hash + kem_ct) pairs
    partial_header = data[:_FIXED_HEADER_SIZE]
    wrap_ad = bytearray(partial_header)
    for i in range(recipient_count):
        offset = stanza_offset + i * _STANZA_SIZE
        wrap_ad.extend(data[offset : offset + _KEM_PK_HASH_SIZE + _KEM_CT_SIZE])
    wrap_ad = bytes(wrap_ad)

    # Unwrap DEK
    dek = aead_decrypt(wrapped_dek, ad=wrap_ad, nonce=_ZERO_NONCE, key=kek)
    if dek is None:
        raise DecryptionError("DEK unwrap failed -- key authentication error")

    # Decrypt data (full header as AD per D-10)
    full_header = data[:expected_header_size]
    ciphertext = data[expected_header_size:]
    plaintext = aead_decrypt(ciphertext, ad=full_header, nonce=data_nonce, key=dek)
    if plaintext is None:
        raise DecryptionError("Data decryption failed -- authentication error")

    # Decompress if suite=0x02 (per D-03, D-09)
    if suite == CIPHER_SUITE_ML_KEM_CHACHA_BROTLI:
        plaintext = _safe_decompress(plaintext, MAX_DECOMPRESSED_SIZE)

    return plaintext


def envelope_parse(data: bytes) -> dict[str, int]:
    """Parse envelope metadata without decryption.

    Returns version, cipher suite, and recipient count from the header.
    Does not require a KEM secret key.

    Args:
        data: Complete or partial binary envelope (at least 8 bytes).

    Returns:
        Dict with keys: version, suite, recipient_count.

    Raises:
        MalformedEnvelopeError: If magic, version, or suite is invalid,
            or data is too short.
    """
    if len(data) < 8:
        raise MalformedEnvelopeError("Envelope too short for header parse")

    magic = data[:4]
    if magic != ENVELOPE_MAGIC:
        raise MalformedEnvelopeError(
            f"Invalid magic: expected CENV, got {magic!r}"
        )

    version = data[4]
    if version != ENVELOPE_VERSION:
        raise MalformedEnvelopeError(f"Unsupported envelope version: {version}")

    suite = data[5]
    if suite not in (CIPHER_SUITE_ML_KEM_CHACHA, CIPHER_SUITE_ML_KEM_CHACHA_BROTLI):
        raise MalformedEnvelopeError(f"Unsupported cipher suite: {suite}")

    recipient_count = struct.unpack(">H", data[6:8])[0]

    return {
        "version": version,
        "suite": suite,
        "recipient_count": recipient_count,
    }

File: cli/src/envelope.cpp
#include "cli/src/envelope.h"
#include "cli/src/wire.h"

#include <oqs/oqs.h>
#include <sodium.h>

#include <algorithm>
#include <array>
#include <cstring>
#include <stdexcept>

namespace chromatindb::cli::envelope {

// =============================================================================
// Constants
// =============================================================================

static constexpr size_t KEM_PK_SIZE       = 1568;
static constexpr size_t KEM_CT_SIZE       = 1568;
static constexpr size_t KEM_SS_SIZE       = 32;
static constexpr size_t KEM_SK_SIZE       = 3168;
static constexpr size_t AEAD_KEY_SIZE     = 32;
static constexpr size_t AEAD_TAG_SIZE     = 16;
static constexpr size_t NONCE_SIZE        = 12;
static constexpr size_t PK_HASH_SIZE      = 32;
static constexpr size_t WRAPPED_DEK_SIZE  = AEAD_KEY_SIZE + AEAD_TAG_SIZE;  // 48
static constexpr size_t STANZA_SIZE       = PK_HASH_SIZE + KEM_CT_SIZE + WRAPPED_DEK_SIZE;  // 1648
static constexpr size_t HEADER_SIZE       = 20;

static constexpr uint8_t MAGIC[4]         = {0x43, 0x45, 0x4E, 0x56};  // "CENV"
static constexpr uint8_t VERSION          = 0x01;
static constexpr uint8_t SUITE            = 0x01;

static constexpr char HKDF_LABEL[] = "chromatindb-envelope-kek-v1";

// =============================================================================
// Internal: SHA3-256 hash
// =============================================================================



// =============================================================================
// Internal: HKDF-SHA256 (extract + expand) via libsodium
// =============================================================================

static std::array<uint8_t, 32> hkdf_derive_kek(
    std::span<const uint8_t> shared_secret) {

    // Extract: PRK = HMAC-SHA256(salt=empty, IKM=shared_secret)
    uint8_t prk[crypto_auth_hmacsha256_BYTES];
    // Empty salt (32 zero bytes for HMAC-SHA256)
    uint8_t salt[crypto_auth_hmacsha256_KEYBYTES] = {};
    crypto_auth_hmacsha256_state extract_state;
    crypto_auth_hmacsha256_init(&extract_state, salt, sizeof(salt));
    crypto_auth_hmacsha256_update(&extract_state, shared_secret.data(), shared_secret.size());
    crypto_auth_hmacsha256_final(&extract_state, prk);

    // Expand: OKM = HMAC-SHA256(PRK, label || 0x01) truncated to 32 bytes
    // For 32 bytes of output, one HMAC round suffices.
    crypto_auth_hmacsha256_state expand_state;
    crypto_auth_hmacsha256_init(&expand_state, prk, sizeof(prk));
    crypto_auth_hmacsha256_update(&expand_state,
        reinterpret_cast<const uint8_t*>(HKDF_LABEL), sizeof(HKDF_LABEL) - 1);
    uint8_t counter = 0x01;
    crypto_auth_hmacsha256_update(&expand_state, &counter, 1);

    std::array<uint8_t, 32> kek{};
    crypto_auth_hmacsha256_final(&expand_state, kek.data());

    sodium_memzero(prk, sizeof(prk));
    return kek;
}

// =============================================================================
// Internal: Stanza data for sorting
// =============================================================================

struct StanzaBuilder {
    std::array<uint8_t, PK_HASH_SIZE> pk_hash;
    std::array<uint8_t, KEM_CT_SIZE> kem_ct;
    std::array<uint8_t, KEM_SS_SIZE> shared_secret;
};

// =============================================================================
// encrypt
// =============================================================================

std::vector<uint8_t> encrypt(
    std::span<const uint8_t> plaintext,
    std::vector<std::span<const uint8_t>> recipient_kem_pubkeys) {

    if (recipient_kem_pubkeys.empty()) {
        throw std::runtime_error("envelope::encrypt: no recipients");
    }
    if (recipient_kem_pubkeys.size() > 65535) {
        throw std::runtime_error("envelope::encrypt: too many recipients");
    }

    // Validate all pubkey sizes
    for (auto& pk : recipient_kem_pubkeys) {
        if (pk.size() != KEM_PK_SIZE) {
            throw std::runtime_error("envelope::encrypt: invalid KEM pubkey size");
        }
    }

    // Generate DEK and data nonce
    std::array<uint8_t, AEAD_KEY_SIZE> dek{};
    randombytes_buf(dek.data(), dek.size());

    std::array<uint8_t, NONCE_SIZE> data_nonce{};
    randombytes_buf(data_nonce.data(), data_nonce.size());

    // Build stanza data: encapsulate for each recipient
    OQS_KEM* kem = OQS_KEM_new(OQS_KEM_alg_ml_kem_1024);
    if (!kem) {
        throw std::runtime_error("envelope::encrypt: failed to create ML-KEM-1024");
    }

    std::vector<StanzaBuilder> stanzas(recipient_kem_pubkeys.size());
    for (size_t i = 0; i < recipient_kem_pubkeys.size(); ++i) {
        // Compute pk_hash = SHA3-256(pubkey)
        stanzas[i].pk_hash = sha3_256(recipient_kem_pubkeys[i]);

        // Encapsulate
        if (OQS_KEM_encaps(kem,
                stanzas[i].kem_ct.data(),
                stanzas[i].shared_secret.data(),
                recipient_kem_pubkeys[i].data()) != OQS_SUCCESS) {
            OQS_KEM_free(kem);
            throw std::runtime_error("envelope::encrypt: ML-KEM encapsulation failed");
        }
    }
    OQS_KEM_free(kem);

    // Sort stanzas by pk_hash (lexicographic)
    std::sort(stanzas.begin(), stanzas.end(),
        [](const StanzaBuilder& a, const StanzaBuilder& b) {
            return a.pk_hash < b.pk_hash;
        });

    // Build fixed header (20 bytes)
    std::array<uint8_t, HEADER_SIZE> header{};
    std::memcpy(header.data(), MAGIC, 4);
    header[4] = VERSION;
    header[5] = SUITE;
    store_u16_be(header.data() + 6, static_cast<uint16_t>(stanzas.size()));
    std::memcpy(header.data() + 8, data_nonce.data(), NONCE_SIZE);

    // Wrap DEK for each stanza with incremental AD
    // AD for stanza[i] wrap = header(20) + accumulated (pk_hash + kem_ct) from stanzas 0..i-1
    static constexpr size_t PK_HASH_PLUS_CT = PK_HASH_SIZE + KEM_CT_SIZE;  // 1600

    // Pre-allocate AD buffer: header + up to N * (pk_hash + kem_ct)
    std::vector<uint8_t> ad_buf;
    ad_buf.reserve(HEADER_SIZE + stanzas.size() * PK_HASH_PLUS_CT);
    ad_buf.insert(ad_buf.end(), header.begin(), header.end());

    std::array<uint8_t, NONCE_SIZE> zero_nonce{};  // all zeros for DEK wrapping

    std::vector<std::array<uint8_t, WRAPPED_DEK_SIZE>> wrapped_deks(stanzas.size());

    for (size_t i = 0; i < stanzas.size(); ++i) {
        // Derive KEK from shared_secret via HKDF
        auto kek = hkdf_derive_kek(stanzas[i].shared_secret);

        // Wrap DEK: encrypt(dek, ad=ad_buf, nonce=zero, key=kek) -> 48 bytes
        unsigned long long ct_len = 0;
        if (crypto_aead_chacha20poly1305_ietf_encrypt(
                wrapped_deks[i].data(), &ct_len,
                dek.data(), AEAD_KEY_SIZE,
                ad_buf.data(), ad_buf.size(),
                nullptr, zero_nonce.data(), kek.data()) != 0) {
            sodium_memzero(dek.data(), dek.size());
            throw std::runtime_error("envelope::encrypt: DEK wrapping failed");
        }

        sodium_memzero(kek.data(), kek.size());
        sodium_memzero(stanzas[i].shared_secret.data(), stanzas[i].shared_secret.size());

        // Accumulate this stanza's pk_hash + kem_ct into AD for next iteration
        ad_buf.insert(ad_buf.end(), stanzas[i].pk_hash.begin(), stanzas[i].pk_hash.end());
        ad_buf.insert(ad_buf.end(), stanzas[i].kem_ct.begin(), stanzas[i].kem_ct.end());
    }

    // AD for data encryption = header(20) + all stanzas(N * 1648)
    std::vector<uint8_t> data_ad;
    data_ad.reserve(HEADER_SIZE + stanzas.size() * STANZA_SIZE);
    data_ad.insert(data_ad.end(), header.begin(), header.end());
    for (size_t i = 0; i < stanzas.size(); ++i) {
        data_ad.insert(data_ad.end(), stanzas[i].pk_hash.begin(), stanzas[i].pk_hash.end());
        data_ad.insert(data_ad.end(), stanzas[i].kem_ct.begin(), stanzas[i].kem_ct.end());
        data_ad.insert(data_ad.end(), wrapped_deks[i].begin(), wrapped_deks[i].end());
    }

    // Encrypt plaintext with DEK
    std::vector<uint8_t> ciphertext(plaintext.size() + AEAD_TAG_SIZE);
    unsigned long long ct_len = 0;
    if (crypto_aead_chacha20poly1305_ietf_encrypt(
            ciphertext.data(), &ct_len,
            plaintext.data(), plaintext.size(),
            data_ad.data(), data_ad.size(),
            nullptr, data_nonce.data(), dek.data()) != 0) {
        sodium_memzero(dek.data(), dek.size());
        throw std::runtime_error("envelope::encrypt: data encryption failed");
    }
    ciphertext.resize(static_cast<size_t>(ct_len));

    sodium_memzero(dek.data(), dek.size());

    // Assemble final envelope
    std::vector<uint8_t> envelope;
    envelope.reserve(HEADER_SIZE + stanzas.size() * STANZA_SIZE + ciphertext.size());

    // Header
    envelope.insert(envelope.end(), header.begin(), header.end());

    // Stanzas
    for (size_t i = 0; i < stanzas.size(); ++i) {
        envelope.insert(envelope.end(), stanzas[i].pk_hash.begin(), stanzas[i].pk_hash.end());
        envelope.insert(envelope.end(), stanzas[i].kem_ct.begin(), stanzas[i].kem_ct.end());
        envelope.insert(envelope.end(), wrapped_deks[i].begin(), wrapped_deks[i].end());
    }

    // Ciphertext + tag
    envelope.insert(envelope.end(), ciphertext.begin(), ciphertext.end());

    return envelope;
}

// =============================================================================
// decrypt
// =============================================================================

std::optional<std::vector<uint8_t>> decrypt(
    std::span<const uint8_t> envelope_data,
    std::span<const uint8_t> our_kem_seckey,
    std::span<const uint8_t> our_kem_pubkey) {

    // Validate minimum size: header(20)
    if (envelope_data.size() < HEADER_SIZE) {
        return std::nullopt;
    }

    // Validate magic
    if (std::memcmp(envelope_data.data(), MAGIC, 4) != 0) {
        return std::nullopt;
    }

    // Validate version and suite
    if (envelope_data[4] != VERSION || envelope_data[5] != SUITE) {
        return std::nullopt;
    }

    // Validate key sizes
    if (our_kem_seckey.size() != KEM_SK_SIZE || our_kem_pubkey.size() != KEM_PK_SIZE) {
        return std::nullopt;
    }

    // Parse header
    uint16_t recipient_count = load_u16_be(envelope_data.data() + 6);
    if (recipient_count == 0) {
        return std::nullopt;
    }

    // Validate total size: header + stanzas + at least the tag
    size_t stanza_total = static_cast<size_t>(recipient_count) * STANZA_SIZE;
    size_t min_size = HEADER_SIZE + stanza_total + AEAD_TAG_SIZE;
    if (envelope_data.size() < min_size) {
        return std::nullopt;
    }

    // data_nonce is at offset 8, 12 bytes
    const uint8_t* data_nonce = envelope_data.data() + 8;

    // Compute our pk_hash
    auto our_pk_hash = sha3_256(our_kem_pubkey);

    // Scan stanzas to find our match
    const uint8_t* stanzas_start = envelope_data.data() + HEADER_SIZE;
    int matched_index = -1;

    for (uint16_t i = 0; i < recipient_count; ++i) {
        const uint8_t* stanza = stanzas_start + static_cast<size_t>(i) * STANZA_SIZE;
        if (std::memcmp(stanza, our_pk_hash.data(), PK_HASH_SIZE) == 0) {
            matched_index = static_cast<int>(i);
            break;
        }
    }

    if (matched_index < 0) {
        return std::nullopt;  // Not a recipient
    }

    // Get our stanza
    const uint8_t* our_stanza = stanzas_start + static_cast<size_t>(matched_index) * STANZA_SIZE;
    const uint8_t* kem_ct = our_stanza + PK_HASH_SIZE;
    const uint8_t* wrapped_dek = kem_ct + KEM_CT_SIZE;

    // Decapsulate
    std::array<uint8_t, KEM_SS_SIZE> shared_secret{};
    OQS_KEM* kem = OQS_KEM_new(OQS_KEM_alg_ml_kem_1024);
    if (!kem) {
        return std::nullopt;
    }

    if (OQS_KEM_decaps(kem, shared_secret.data(), kem_ct, our_kem_seckey.data()) != OQS_SUCCESS) {
        OQS_KEM_free(kem);
        return std::nullopt;
    }
    OQS_KEM_free(kem);

    // Derive KEK
    auto kek = hkdf_derive_kek(shared_secret);
    sodium_memzero(shared_secret.data(), shared_secret.size());

    // Rebuild AD for this stanza's position:
    // header(20) + accumulated (pk_hash + kem_ct) from stanzas 0..matched_index-1
    static constexpr size_t PK_HASH_PLUS_CT = PK_HASH_SIZE + KEM_CT_SIZE;
    std::vector<uint8_t> wrap_ad;
    wrap_ad.reserve(HEADER_SIZE + static_cast<size_t>(matched_index) * PK_HASH_PLUS_CT);
    wrap_ad.insert(wrap_ad.end(), envelope_data.data(), envelope_data.data() + HEADER_SIZE);

    for (int i = 0; i < matched_index; ++i) {
        const uint8_t* s = stanzas_start + static_cast<size_t>(i) * STANZA_SIZE;
        // pk_hash(32) + kem_ct(1568) = 1600 bytes
        wrap_ad.insert(wrap_ad.end(), s, s + PK_HASH_PLUS_CT);
    }

    // Unwrap DEK
    std::array<uint8_t, AEAD_KEY_SIZE> dek{};
    unsigned long long dek_len = 0;
    std::array<uint8_t, NONCE_SIZE> zero_nonce{};

    if (crypto_aead_chacha20poly1305_ietf_decrypt(
            dek.data(), &dek_len,
            nullptr,
            wrapped_dek, WRAPPED_DEK_SIZE,
            wrap_ad.data(), wrap_ad.size(),
            zero_nonce.data(), kek.data()) != 0) {
        sodium_memzero(kek.data(), kek.size());
        return std::nullopt;
    }
    sodium_memzero(kek.data(), kek.size());

    // Build AD for data decryption: header(20) + all stanzas(N * 1648)
    std::vector<uint8_t> data_ad;
    data_ad.reserve(HEADER_SIZE + stanza_total);
    data_ad.insert(data_ad.end(), envelope_data.data(), envelope_data.data() + HEADER_SIZE);
    data_ad.insert(data_ad.end(), stanzas_start, stanzas_start + stanza_total);

    // Decrypt ciphertext
    const uint8_t* ciphertext = stanzas_start + stanza_total;
    size_t ciphertext_len = envelope_data.size() - HEADER_SIZE - stanza_total;

    if (ciphertext_len < AEAD_TAG_SIZE) {
        sodium_memzero(dek.data(), dek.size());
        return std::nullopt;
    }

    std::vector<uint8_t> decrypted(ciphertext_len - AEAD_TAG_SIZE);
    unsigned long long pt_len = 0;

    if (crypto_aead_chacha20poly1305_ietf_decrypt(
            decrypted.data(), &pt_len,
            nullptr,
            ciphertext, ciphertext_len,
            data_ad.data(), data_ad.size(),
            data_nonce, dek.data()) != 0) {
        sodium_memzero(dek.data(), dek.size());
        return std::nullopt;
    }
    sodium_memzero(dek.data(), dek.size());

    decrypted.resize(static_cast<size_t>(pt_len));
    return decrypted;
}

// =============================================================================
// is_envelope
// =============================================================================

bool is_envelope(std::span<const uint8_t> data) {
    if (data.size() < 4) {
        return false;
    }
    return std::memcmp(data.data(), MAGIC, 4) == 0;
}

} // namespace chromatindb::cli::envelope
